upload.php

```diff
@@ -0,0 +1,135 @@
+$step = optional_param('step', '', PARAM_ALPHA); // The current step in the process.
+
+if (!$facetoface = $DB->get_record('facetoface', ['id' => $f])) {
+    throw new moodle_exception('error:incorrectfacetofaceid', 'facetoface');
+}
+if (!$course = $DB->get_record('course', ['id' => $facetoface->course])) {
+    throw new moodle_exception('error:coursemisconfigured', 'facetoface');
+}
+if (!$cm = get_coursemodule_from_instance('facetoface', $facetoface->id, $course->id)) {
+    throw new moodle_exception('error:incorrectcoursemoduleid', 'facetoface');
+}
+
+
+require_course_login($course, true, $cm);
+$context = context_course::instance($course->id);
+$modulecontext = context_module::instance($cm->id);
+require_capability('mod/facetoface:editsessions', $context);
+require_capability('mod/facetoface:uploadbookings', $context);
+
+
+echo $OUTPUT->header();
+
+// Render form, which should only consist of an upload element.
+if ($validate) {
+    // Form submitted, but not ready for processing -> validate.
+    $heading = get_string('facetoface:validatebookings', 'facetoface');
+
+    $mform = new upload_bookings_form(null);
+    $data = $mform->get_data();
+    $fileid = $data->csvfile ?: 0;
+
+    $mform = new confirm_bookings_form(null, ['f' => $f, 'fileid' => $fileid]);
+
+    $bm = new booking_manager($f);
+    $bm->load_from_file($fileid);
+
+    // Validate entries.
+    $errors = $bm->validate();
+
+    // List out any issues in a table.
+    if (!empty($errors)) {
+        // Print summary statement.
+        echo \core\notification::error(get_string('error:bookingsuploadfileerrorsfound', 'mod_facetoface', count($errors)));
+
+        $table = new html_table();
+        $table->attributes['class'] = 'f2fbookingsuploadlist m-auto generaltable mb-2';
+
+        $table->head[] = get_string('uucsvline', 'tool_uploaduser');
+        $table->head[] = get_string('status');
+        $table->data = $errors;
+
+        echo html_writer::tag('div', html_writer::table($table), ['class' => 'flexible-wrap mb-4']);
+    } else {
+        // Bonus: show a preview/summary for good records (e.g. 40 records will be processed).
+        echo \core\notification::success(get_string('facetoface:uploadreadytoprocess', 'mod_facetoface'));
+    }
+
+
+    // Set form data to allow user to continue and process the uploaded file on their next form submit.
+} else if ($process && $fileid && $f) {
+    // Form submitted, and ready for processing -> process.
+    $bm = new booking_manager($f);
+    $bm->load_from_file($fileid);
+
+    // Validate entries.
+    $errors = $bm->validate();
+    if (empty($errors)) {
+        // Process entries.
+        $bm->process();
+
+        echo "PROCESSED";
+        die;
+        // Redirect back to start with notification.
+        redirect(
+            new moodle_url('/mod/facetoface/upload.php', ['f' => $f]),
+            'Successfully processed X records.',
```

Comment:
This bit looks like it wants to be a parameterised lang string?

Reply:
I made it a bit more general. The number of records processed would equal the number of rows in the file. I don't think that information is important as it's either all or nothing, as it wouldn't redirect if there was an exception during processing.